### ISSUE TITLE
Bean-price: support fetch over range of dates

### ISSUE THREAD

**Issue by blais** [beancount/beanprice#2]
**[Original report](https://bitbucket.org/blais/beancount/issue/329) by Vivek G (Bitbucket: [seltzered](https://bitbucket.org/seltzered), GitHub: [seltzered](https://github.com/seltzered)).**

----------------------------------------

I've added initial support for bean-price to fetch over a range of dates in https://bitbucket.org/seltzered/beancount/commits/branch/bean-price-daterange . 

Currently it attempts to fetch over dates that may be invalid (e.g. weekends) and will filter out redundant results.

**Comment by AshwinParanjape** [comment]
I wanted to confirm - currently the script has an `--update` argument but it throws the following error 
```
Traceback (most recent call last):
  File "/opt/anaconda3/envs/FinMan/bin/bean-price", line 8, in <module>
    sys.exit(main())
  File "/opt/anaconda3/envs/FinMan/lib/python3.7/site-packages/beanprice/price.py", line 679, in main
    args, jobs, entries, dcontext = process_args()
  File "/opt/anaconda3/envs/FinMan/lib/python3.7/site-packages/beanprice/price.py", line 652, in process_args
    jobs.extend(get_price_jobs_up_to_date(entries,
NameError: name 'get_price_jobs_up_to_date' is not defined
```

Looking at the code, I found no function by the name `get_price_jobs_up_to_date`, so is that because the function got lost when bean-price was separated out of beancount repo, or is it because the function was never written?

**Comment by blais** [comment]
The beanprice repo is sync'ed to beancount branch 'master'

**Comment by xuhcc** [comment]
>The beanprice repo is sync'ed to beancount branch 'master'

@blais What do you mean by that?

`beanprice/price.py` has undefined variable `get_price_jobs_up_to_date` at line 652. It was introduced in commit  (f1bcfea1c217c460ef419e05aa7b9ab2dff0bab2).

**Comment by blais** [comment]
Hmm, looks like a git fart from my part. I merged pr128_seltzered_beanpriceupdate by hand because it was impossible to update after the changes, probably missed the function somehow? Tests run, but don't cover that case.
I'll try to dig it out

**Comment by blais** [comment]
Found it (thank good for an old clone, I found the original branch ref and it wasn't GCed).
Adding it. There's still no test for it. I ran it, gathers jobs (but then something changed in Yahoo,response fails on missing timestamp).

Lessons learned: Never delete refs for branches not fully merged; archive to tags instead.

What I meant with my comment is that beanprice (and all other repos I'm extracting from it eventually) is assumed to work against beancount @ master, not its v2 branch, which is self-contained (I made the wrong assumption in my quick response from mobile phone because this was a fresh commit).

**Comment by blais** [comment]
https://github.com/beancount/beanprice/commit/b2685fea866a8696f9d5c25e851d30b62d6a02a2

**Comment by xuhcc** [comment]
>What I meant with my comment is that beanprice (and all other repos I'm extracting from it eventually) is assumed to work against beancount @ master, not its v2 branch, which is self-contained

`beanprice` works with beancount v2, that's how I'm using it. There's no reason to break backwards compatibilty.

>b2685fe

I haven't actually run the code, but pylint reports invalid reference:
```
beanprice/price.py:386: [E0601(used-before-assignment), get_price_jobs_up_to_date] Using variable 'base' before assignment
```

**Comment by blais** [comment]
I'm not sure that the combo of beanprice repo + beancount v2 still works (no guarantees).
If you want to use bean-price with v2, I would suggest using the bean-price that's inside v2.

If we want to support beanprice repo + v2 I think it would be best to remove bean-price from v2, but I sort-of decided not to work on v2 anymore (lest it cause more work and time sink). But the problem remains: doing the right thing for beanprice sometimes requires adding functionality in beancount itself (e.g. generic functions) and I'd rather do that only in v3 going forward.

**Comment by xuhcc** [comment]
>I'm not sure that the combo of beanprice repo + beancount v2 still works (no guarantees).

As I said, beanprice (version 1.0.1 specifically) works fine with beancount v2 (v2.3.3 in my case). The tests also pass with latest master. The new `--update` flag may not work in some cases regardless of beancount version because the code uses undeclared variable `base` at line 386. I think it would be better to revert these changes and submit them properly via pull request.

>If we want to support beanprice repo + v2 I think it would be best to remove bean-price from v2, but I sort-of decided not to work on v2 anymore (lest it cause more work and time sink).

beancount v2 is stable and a lot of people using it. There is no need to remove anything from it.

beanprice is a standalone package for those who want extra features or additional price sources.

>But the problem remains: doing the right thing for beanprice sometimes requires adding functionality in beancount itself (e.g. generic functions) and I'd rather do that only in v3 going forward.

If there will be such functionality, it could be simply turned off for beancount v2.

**Comment by blais** [comment]
> > I'm not sure that the combo of beanprice repo + beancount v2 still works (no guarantees).
> 
> As I said, beanprice (version 1.0.1 specifically) works fine with beancount v2 (v2.3.3 in my case). The tests also pass with latest master.

There's a misunderstanding. It *happens* to work. Whether we *want* to support that combo working is the question.
To keep that combo working, it means we have to commit occasional changes on branch v2.
I don't have the cycles, and besides, I want to let v2 gently rot with no action so everyone moves over eventually (by then master should be ready). I you send beancount repo PRs on branch v2 for bean-price changes I could review them, but I'd like to prioritize 99% of time on new development on v3 instead.

So I suggest instead that at some point we just state bean-price + master is the only supported combo.
It's also a simpler story to understand: 
- v2 is self-contained; 
- v3 is spread over multiple repositories which are intended to be in sync at branch master.


> The new `--update` flag may not work in some cases regardless of beancount version because the code uses undeclared variable `base` at line 386. I think it would be better to revert these changes and submit them properly via pull request.

I'd rather we write a unit test for that branch (that's from the original PR) and just fix the bug. Less time.
But you're welcome to revert and resubmit it you prefer.


> 
> > If we want to support beanprice repo + v2 I think it would be best to remove bean-price from v2, but I sort-of decided not to work on v2 anymore (lest it cause more work and time sink).
> 
> beancount v2 is stable and a lot of people using it. There is no need to remove anything from it.
> 
> beanprice is a standalone package for those who want extra features or additional price sources.

So in your mind it's a v2-compatible package?
Is it supposed to work with v3?

> 
> > But the problem remains: doing the right thing for beanprice sometimes requires adding functionality in beancount itself (e.g. generic functions) and I'd rather do that only in v3 going forward.
> 
> If there will be such functionality, it could be simply turned off for beancount v2.

**Comment by xuhcc** [comment]
>So in your mind it's a v2-compatible package?
>Is it supposed to work with v3?

It is supposed to work with both versions. You also suggested that it could work with ledger and hledger: https://groups.google.com/g/beancount/c/zSqZOM0a7Jk

**Comment by blais** [comment]
Let's just see how it goes. You're welcome to make any changes to dispatch to both if you want to keep the bean-price + v2 combo working (e.g. disable some features in bean-price if it is backed by v2).  I do want to make sure it works with master though.

Ledger and HLedger are a bit different; there's no code dependency on them, so it's free so far (just output formatting). I suppose a nice generalization project would be to support figuring out the list of jobs to make from Ledger or HLedger files, and then format accordingly, leveraging the common code to fetch from price sources.

**Comment by blais** [comment]
How about this: We should isolate the Beancount code to just a few modules, and turn the scope of this project larger: it's three things:
a) a Python library of price fetchers up-to-date with unit tests and a common good (needs to be improved from what's there TBH) interface;
b) an accompanying tool for fetching those prices from the command-line
c) OPTIONAL interfaces to fetch ledger information (for missing/required prices) onto beancount v2, beancount v3, ledger, hledger, and rendering output formats to any of these.

In other words, make much of the "bean" from beanprice as a side-feature. Turn this project into something that can run independent of beancount.

I'll send some CLs, hopefully I can find some time during the holidays.

**Comment by xuhcc** [comment]
I like this idea but I want `beanprice` to remain compatible with existing beancount price sources for a long period of time.

>CLs

What's that?

**Comment by blais** [comment]
> I like this idea but I want `beanprice` to remain compatible with existing beancount price sources for a long period of time.

Oh yes, there's no particular reason to break any of these.


> 
> > CLs
> 
> What's that?

Sorry. Google terminology. Stands for "change list."
